src/compiler/checker.ts

```diff
@@ -41246,23 +41181,12 @@ namespace ts {
             if (!parameter.type) {
                 return grammarErrorOnNode(parameter.name, Diagnostics.An_index_signature_parameter_must_have_a_type_annotation);
             }
-            if (parameter.type.kind !== SyntaxKind.StringKeyword && parameter.type.kind !== SyntaxKind.NumberKeyword) {
-                const type = getTypeFromTypeNode(parameter.type);
-
-                if (type.flags & TypeFlags.String || type.flags & TypeFlags.Number) {
-                    return grammarErrorOnNode(parameter.name,
-                        Diagnostics.An_index_signature_parameter_type_cannot_be_a_type_alias_Consider_writing_0_Colon_1_Colon_2_instead,
-                        getTextOfNode(parameter.name),
-                        typeToString(type),
-                        typeToString(node.type ? getTypeFromTypeNode(node.type) : anyType));
-                }
-
-                if (type.flags & TypeFlags.Union && allTypesAssignableToKind(type, TypeFlags.StringOrNumberLiteral, /*strict*/ true)) {
-                    return grammarErrorOnNode(parameter.name,
-                        Diagnostics.An_index_signature_parameter_type_cannot_be_a_union_type_Consider_using_a_mapped_object_type_instead);
-                }
-
-                return grammarErrorOnNode(parameter.name, Diagnostics.An_index_signature_parameter_type_must_be_either_string_or_number);
+            const type = getTypeFromTypeNode(parameter.type);
+            if (someType(type, t => !!(t.flags & (TypeFlags.StringOrNumberLiteralOrUnique | TypeFlags.Instantiable)) && !isPatternLiteralType(t))) {
+                return grammarErrorOnNode(parameter.name, Diagnostics.An_index_signature_parameter_type_cannot_be_a_literal_type_or_generic_type_Consider_using_a_mapped_object_type_instead);
+            }
+            if (!everyType(type, t => !!(t.flags & (TypeFlags.String | TypeFlags.Number | TypeFlags.ESSymbol | TypeFlags.TemplateLiteral)))) {

```

Comment:
`TypeFlags.StringMapping`?

Reply:
Nope, `TypeFlags.StringMapping` is an instantiable type and it was already barred by the test above.

Reply:
Hm, on inspection,

```ts
type A = Uppercase<`${number}`>;
```

Is just

```ts
type A = `${number}`;
```

so they just evaporate around patterns (after affecting the concrete bits, even though, conceptually, they should represent a different, smaller, set of strings in the holes), so I guess we don't need to handle them right now. But if we ever change up mappings over/in patterns to be preserved, we'll need to change this to match, I guess.

Reply:
Yeah, we're good there for now. However, it *does* appear we don't properly apply the string mapping to template literal types with non-generic patterns. For example, `` Uppercase<`foo${string}bar`> `` should become `` `FOO${string}BAR` ``, but simply resolves to `` `foo${string}bar` `` now. It's unrelated to this PR, but something we should fix.